File: baekjoon_all/01000+/boj_1260.java
// Solve 2024-02-18

// 백준에 제출할 때는 class 이름을 Main으로 설정해야 한다.

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Collections;
import java.util.StringTokenizer;

public class boj_1260 {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        int startNode = Integer.parseInt(st.nextToken());

        ArrayList<Integer>[] adj = new ArrayList[n + 1];

        for (int u = 1; u <= n; u++) {
            adj[u] = new ArrayList<Integer>();
        }

        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine(), " ");
            int from = Integer.parseInt(st.nextToken());
            int to = Integer.parseInt(st.nextToken());
            adj[from].add(to);
            adj[to].add(from);
        }

        for (int u = 1; u <= n; u++) {
            Collections.sort(adj[u]);
        }

        boolean[] visited = new boolean[n + 1];
        ArrayDeque<int[]> dfsStack = new ArrayDeque<int[]>();
        visited[startNode] = true;
        sb.append(startNode).append(" ");
        dfsStack.offerLast(new int[] { startNode, 0 });

        while (!dfsStack.isEmpty()) {
            int[] cur = dfsStack.pollLast();

            for (int i = cur[1]; i < adj[cur[0]].size(); i++) {
                int nextNode = adj[cur[0]].get(i);

                if (!visited[nextNode]) {
                    visited[nextNode] = true;
                    sb.append(nextNode).append(" ");
                    dfsStack.add(new int[] { cur[0], i + 1 });
                    dfsStack.add(new int[] { nextNode, 0 });
                    break;
                }
            }
        }

        sb.append("\n");

        for (int u = 1; u <= n; u++) {
            visited[u] = false;
        }

        ArrayDeque<Integer> bfsQueue = new ArrayDeque<Integer>();
        visited[startNode] = true;
        bfsQueue.offerLast(startNode);

        while (!bfsQueue.isEmpty()) {
            int curNode = bfsQueue.pollFirst();
            sb.append(curNode).append(" ");

            for (int nextNode : adj[curNode]) {
                if (!visited[nextNode]) {
                    visited[nextNode] = true;
                    bfsQueue.offerLast(nextNode);
                }
            }
        }

        sb.append("\n");
        System.out.print(sb);
        br.close();
    }

}

File: baekjoon_all/01000+/boj_1260.py
# Solve 2025-09-07
# Update 2025-09-22

import sys
from collections import deque

input = lambda: sys.stdin.readline().rstrip()

def dfs(cur, adj, visited, dfs_output):
    visited[cur] = True
    dfs_output.append(cur)

    for nxt in adj[cur]:
        if not visited[nxt]:
            dfs(nxt, adj, visited, dfs_output)


def bfs(v, adj, visited, bfs_output):
    que = deque()
    visited[v] = True
    que.append(v)
    bfs_output.append(v)

    while que:
        cur = que.popleft()

        for nxt in adj[cur]:
            if not visited[nxt]:
                visited[nxt] = True
                que.append(nxt)
                bfs_output.append(nxt)


n, m, v = map(int, input().split())
adj = [[] for _ in range(n + 1)]

for _ in range(m):
    u1, u2 = map(int, input().split())
    adj[u1].append(u2)
    adj[u2].append(u1)

for li in adj:
    li.sort()

visited = [False] * (n + 1)
dfs_output = []
dfs(v, adj, visited, dfs_output)
print(*dfs_output)

visited = [False] * (n + 1)
bfs_output = []
bfs(v, adj, visited, bfs_output)
print(*bfs_output)
